stratum/hal/lib/common/hal.cc

```diff
@@ -163,7 +187,33 @@ ::util::Status Hal::Teardown() {
   // of warmboot shutdown, the stack is first freezed by calling an RPC in
   // AdminService, which itself calls Freeze() method in SwitchInterface class.
   LOG(INFO) << "Shutting down HAL...";
+
   ::util::Status status = ::util::OkStatus();
+
+  if (external_server_shutdown_thread_) {
+    void* value;
+    int ret = pthread_join(external_server_shutdown_thread_, &value);
+    if (ret != 0) {
+      ::util::Status error =
+          MAKE_ERROR(ERR_INTERNAL)
+          << "External server shutdown thread could not be joined. "
+          << "Error: " << strerror(ret) << " (" << ret << ")";
+      APPEND_STATUS_IF_ERROR(status, error);
+    }
+    if (value == PTHREAD_CANCELED) {
+      LOG(ERROR) << "External server shutdown thread was canceled. This is "
+                    "okay in a test, but should otherwise be avoided.";
+    } else if (value != nullptr) {
+      ::util::Status error =
+          MAKE_ERROR(ERR_INTERNAL)
+          << "External server shutdown thread returned unexpected value.";
+      APPEND_STATUS_IF_ERROR(status, error);
```

Comment:
This branch is probably not necessary as nullptr is literally the only thing we return.

Reply:
I feel the same about the other checks too.

Instead of trying to expect everything, focus on the single one implemented path. Any deviation is an error.

What is the user supposed to do with a log message like this: `External server shutdown thread was canceled. This is okay in a test, but should otherwise be avoided.` ?